finality-verifiers/grandpa/src/side_effects.rs

```diff
@@ -0,0 +1,281 @@
+use crate::{Config, Error};
+use codec::{Decode, Encode};
+use sp_runtime::DispatchError;
+use sp_std::{vec, vec::Vec};
+
+#[derive(Encode, Decode)]
+pub enum TransferEventStub<T: frame_system::Config, Balance> {
+    Endowed(T::AccountId, Balance),
+    DustLost(T::AccountId, Balance),
+    Transfer {
+        from: T::AccountId,
+        to: T::AccountId,
+        amount: Balance,
+    },
+}
+
+type CurrencyId = u32;
+
+#[derive(Encode, Decode)]
+pub enum MultiTransferEventStub<T: frame_system::Config, Balance, CurrencyId> {
+    Endowed(CurrencyId, T::AccountId, Balance),
+    DustLost(CurrencyId, T::AccountId, Balance),
+    Transfer {
+        currency_id: CurrencyId,
+        from: T::AccountId,
+        to: T::AccountId,
+        amount: Balance,
+    },
+}
+
+pub(crate) fn decode_event<T: Config<I>, I: 'static>(
+    id: &[u8; 4],
+    mut encoded_event: Vec<u8>,
+    value_abi_unsigned_type: &[u8],
+) -> Result<(Vec<Vec<u8>>, Vec<u8>), DispatchError> {
+    // the first byte is the pallet index, which we don't need
+    let _ = encoded_event.remove(0);
+    match &id {
+        &b"tran" => {
```

Comment:
Do these need to be tied in with types::SIDE_EFFECT_KIND or smth similar to above?

Reply:
I wanted this pallet to not depend on any circuit types. Ideally going forward we define a `FinalityVerifier` trait that includes all of these things and ensures every FV is compatible.